boot/bootstate20.go

```diff
@@ -21,330 +21,526 @@ package boot
 
 import (
 	"fmt"
+	"path/filepath"
 
 	"github.com/snapcore/snapd/bootloader"
 	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
 	"github.com/snapcore/snapd/snap"
 )
 
-type bootState20Base struct{}
+// bootSnap is an interface that is used to make the generic methods, setNext()
+// and markSuccessful() easier to implement
+type bootSnap interface {
```

Comment:
I still think we should try not to need this,  what if:

* we change revisions to return full status and not just trying bool ?  (we need to double check things are consistent and error there otherwise though)
* we make genericSetNext more functional and return  trySnap or nil  and what to do with status  (this would stop us needing setStatus and setTrySnap), see the issue with spurious writes though

Reply:
see my other comment about erroring vs not in revisions

Reply:
I will try that out, I had forgotten about changing this interface

Reply:
I implemented this, along with only writing when something has changed for the kernel_status.